extensions/ql-vscode/src/common/vscode/multi-cancellation-token.ts

```diff
@@ -0,0 +1,28 @@
+import { CancellationToken, Disposable } from "vscode";
+
+/**
+ * A cancellation token that cancels when any of its constituent
+ * cancellation tokens are cancelled.
+ */
+export class MultiCancellationToken implements CancellationToken {
+  private readonly tokens: CancellationToken[];
+
+  constructor(...tokens: CancellationToken[]) {
+    this.tokens = tokens;
+  }
+
+  get isCancellationRequested(): boolean {
+    return this.tokens.some((t) => t.isCancellationRequested);
+  }
+
+  onCancellationRequested<T>(listener: (e: T) => any): Disposable {
```

Comment:
I'm not sure we're implementing `onCancellationRequested` here in line with how it's intended. Maybe this doesn't matter in practice for this usage, but since we're writing a reusable piece of plumbing here it would be nice to do implement it as fully as possible.

As I understand it, the purpose of `onCancellationRequested` is to be an event that code can listen to and know when the token has been cancelled. Therefore rather than us calling `onCancellationRequested` on all the child tokens, I believe the flow of information/control should be the other way around and we should instead be listening for one of the child tokens to fire its `onCancellationRequested` event.

This requires changes to a few different areas of the class so I can't really do this as a "suggestion" block, but I think the code would look like:

```typescript
export class MultiCancellationToken implements CancellationToken {
  private readonly tokens: CancellationToken[];
  private readonly onCancellationRequestedEvent = new EventEmitter<void>();

  constructor(...tokens: CancellationToken[]) {
    this.tokens = tokens;
    tokens.forEach((t) =>
      t.onCancellationRequested(() => this.onCancellationRequestedEvent.fire()),
    );
  }

  get isCancellationRequested(): boolean {
    return this.tokens.some((t) => t.isCancellationRequested);
  }

  get onCancellationRequested(): Event<any> {
    return this.onCancellationRequestedEvent.event;
  }
}
```

What do you think?

Reply:
Thanks. Good suggestion.

Regarding cancellation by mouse hover, the way caching works, the first time the query runs to completion, all other query calls will use the cached results instead of calling it again. Sometimes the query completes quickly so it's hard to get the hover to cancel.

Can you try this on a larger file in the archive?

Reply:
> the first time the query runs to completion, all other query calls will use the cached results instead of calling it again

This is the behaviour I'm seeing, but unfortunately I still can't get it to cancel by breaking the hover. I picked a large file so the query took a few seconds to run. When it was running I could cancel the query by using the progress bar notification, but moving the mouse around seemingly did nothing. Once it had run to completion once it didn't run again and used cached results.

I tried adding in some logging to `MultiCancellationToken.isCancellationRequested` and it seems that neither of the sub-tokens was considered cancelled. So there's not anything wrong with `MultiCancellationToken` but there is with the way I'm testing because VS Code isn't cancelling the token when I move the mouse. Could there perhaps be a setting that controls this, so our VS Code instances are behaving differently?

As I said above, the code looks fine to me, so if we can't understand it and it works on your machine I'm happy to believe it and merge the code.

Reply:
I will verify this again locally. I'll also try to get a second screen recording.